samplers/parser.go

```diff
@@ -84,6 +85,75 @@ func ValidMetric(sample *UDPMetric) bool {
 	return ret
 }
 
+const expectedFrameVersion uint32 = 0
+const frameHeaderLength = 10
+const maxPacketLength = 16 * 1024 * 1024 // 16MB ought to be enough for anyone (for now)
+
+var (
+	// ErrInvalidFrameLen is an error indicating a malformed frame
+	// on an SSF message stream. All headers are expected to be 10 bytes.
+	ErrInvalidFrameLen = fmt.Errorf("Protocol violation: SSFFrame is not the correct length")
+)
+
+// SSFStreamHeader constructs and returns an SSF frame header for the
+// byte sequence buf.
+func SSFStreamHeader(buf []byte) *ssf.SSFFrameHeader {
+	length := uint32(len(buf))
+	version := expectedFrameVersion
+	return &ssf.SSFFrameHeader{
+		Version: &version,
+		Length:  &length,
+	}
+}
+
+// ReadSSFStream takes a reader and returns an array of unvalidated
+// SSF spans. Callers of ParseSSFStream have to take care to validate
+// the traces contained in each span and to extract any metrics from
+// them.
+func ReadSSFStream(reader io.Reader) ([]*ssf.SSFSpan, error) {
+	frameHdr := ssf.SSFFrameHeader{}
+	// TODO: This probably wants to be a pool
+	header := make([]byte, frameHeaderLength)
+	n, err := reader.Read(header)
+	if err != nil {
+		return []*ssf.SSFSpan{}, err
+	}
+	if n != frameHeaderLength {
+		return []*ssf.SSFSpan{}, ErrInvalidFrameLen
+	}
+	// TODO: This scratchBuff thing feels really dirty. This too wants to be a pool, I'm sure.
```

Comment:
Yeah, a pool would become worthwhile if we separate the unmarshaling into a worker. Without that we can't get any concurrency here since a single reader needs to delimit the stream, right?